crates/next-dev/benches/util/mod.rs

```diff
@@ -131,105 +132,131 @@ pub fn resume_on_error<F: FnOnce() + UnwindSafe>(f: F) {
     }
 }
 
-pub trait AsyncBencherExtension {
-    fn try_iter_async<I, O, S, SF, W, WF, R, F, T, TF>(
+pub trait AsyncBencherExtension<A: AsyncExecutor> {
+    fn try_iter_async<I, O, S, SF, R, F, U, UF, T, TF>(
         &mut self,
+        runner: A,
         setup: S,
-        warmup: W,
         routine: R,
+        restore: U,
         teardown: T,
     ) where
         S: Fn() -> SF,
         SF: Future<Output = Result<I>>,
-        W: Fn(I) -> WF,
-        WF: Future<Output = Result<I>>,
         R: Fn(I) -> F,
         F: Future<Output = Result<O>>,
+        U: Fn(O) -> UF,
+        UF: Future<Output = Result<I>>,
         T: Fn(O) -> TF,
         TF: Future<Output = ()>;
 }
 
-impl<'a, 'b, A: AsyncExecutor> AsyncBencherExtension for AsyncBencher<'a, 'b, A, WallTime> {
+impl<'a, 'b, A: AsyncExecutor> AsyncBencherExtension<A> for AsyncBencher<'a, 'b, A, WallTime> {
     #[inline(never)]
-    fn try_iter_async<I, O, S, SF, W, WF, R, F, T, TF>(
+    fn try_iter_async<I, O, S, SF, R, F, U, UF, T, TF>(
         &mut self,
+        runner: A,
         setup: S,
-        warmup: W,
         routine: R,
+        restore: U,
         teardown: T,
     ) where
         S: Fn() -> SF,
         SF: Future<Output = Result<I>>,
-        W: Fn(I) -> WF,
-        WF: Future<Output = Result<I>>,
         R: Fn(I) -> F,
         F: Future<Output = Result<O>>,
+        U: Fn(O) -> UF,
+        UF: Future<Output = Result<I>>,
         T: Fn(O) -> TF,
         TF: Future<Output = ()>,
     {
-        let config = std::env::var("TURBOPACK_BENCH_BENCH").ok();
-        let bench_benchmark_itself = !matches!(
-            config.as_deref(),
-            None | Some("") | Some("no") | Some("false")
-        );
         let log_progress = !matches!(
             std::env::var("TURBOPACK_BENCH_PROGRESS").ok().as_deref(),
             None | Some("") | Some("no") | Some("false")
         );
+        let log_progress_verbose = matches!(
+            std::env::var("TURBOPACK_BENCH_PROGRESS").ok().as_deref(),
+            Some("verbose")
+        );
 
         let setup = &setup;
-        let warmup = &warmup;
         let routine = &routine;
+        let restore = &restore;
         let teardown = &teardown;
+        let input = &Mutex::new(Some(black_box(runner.block_on(async {
+            let measurement = WallTime;
+            if log_progress_verbose {
+                eprint!(" setup...");
+            }
+            let start = measurement.start();
+            let input = retry_async_default((), |_| setup())
+                .await
+                .expect("failed to setup");
+            if log_progress_verbose {
+                let duration = measurement.end(start);
+                eprint!(" [{}]", FormatDuration(duration));
```

Comment:
`FormatDuration` formats `12034ms` as `12s`. Use the `Debug` impl of `Duration` instead which is more precise.

Reply:
Why do you need it that precise?

Reply:
It's useful to see possibly small changes quickly.